libc/mman.c

```diff
@@ -0,0 +1,362 @@
+    int fd,
+    off_t offset)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    int flags_copy = flags;
+
+    oe_errno = OE_EINVAL;
+
+    // If addr is not NULL, then the kernel takes it as a hint about where to
+    // place the mapping; on Linux, the kernel will pick a nearby page boundary
+    // (but always above or equal to the value specified by
+    // /proc/sys/vm/mmap_min_addr) and attempt to create the mapping there.
+    // OE currently does not support this usage.
+    if (addr != NULL)
+        OE_RAISE_MSG(
+            OE_UNSUPPORTED, "non-zero `addr` parameter is not supported.");
+
+    // PROT_NONE and PROT_EXEC are not supported.
+    if (prot == PROT_NONE || (prot & PROT_EXEC))
+        OE_RAISE_MSG(OE_UNSUPPORTED, "unsupported `prot` value %d", prot);
+
+    enum
+    {
+        UNSUPPORTED,
+        IGNORED,
+        SUPPORTED
+    };
+    static struct support
+    {
+        int flag;
+        int support;
+    } flags_table[] = {
+        {MAP_SHARED, SUPPORTED},
+        {MAP_SHARED_VALIDATE, SUPPORTED},
+        {MAP_PRIVATE, SUPPORTED},
+#ifdef MAP_32BIT
+        {MAP_32BIT, UNSUPPORTED},
+#endif
+        {MAP_ANON, SUPPORTED},
+        {MAP_ANONYMOUS, SUPPORTED},
+        {MAP_DENYWRITE, IGNORED /* by spec */},
+        {MAP_EXECUTABLE, IGNORED /* by spec */},
+        {MAP_FILE, IGNORED /* by spec */},
+        {MAP_FIXED, UNSUPPORTED},
+        {MAP_FIXED_NOREPLACE, UNSUPPORTED},
+        {MAP_GROWSDOWN, UNSUPPORTED},
+        {MAP_HUGETLB, UNSUPPORTED},
+        {MAP_HUGE_2MB, UNSUPPORTED},
+        {MAP_HUGE_1GB, UNSUPPORTED},
+        {MAP_LOCKED, UNSUPPORTED},
+        {MAP_NONBLOCK, IGNORED /* no special handling by OE */},
+        {MAP_NORESERVE, IGNORED /* no special handling by OE */},
+        {MAP_POPULATE, IGNORED /* no special handling by OE */},
+        {MAP_STACK, IGNORED /* currently no-op on Linux */},
+        {MAP_SYNC, IGNORED /* no special handling needed for OE */},
+        // MUSL doesn't defined MAP_UNINITIALIZED
+        // { MAP_UNINITIALIZED, SUPPORTED }
+    };
+
+    // Of the above flags, only MAP_FIXED is specified in POSIX.1-2001 and
+    // POSIX.1-2008.  However, most systems also support MAP_ANONYMOUS.
+
+    for (int i = 0; i < OE_COUNTOF(flags_table); ++i)
+    {
+        if (flags_copy & flags_table[i].flag)
+        {
+            if (flags_table[i].support == UNSUPPORTED)
+                OE_RAISE_MSG(
+                    OE_UNSUPPORTED,
+                    "unsupported `flag` value %d",
+                    flags_table[i].flag);
+
+            // Remove flag.
+            flags_copy &= ~flags_table[i].flag;
+        }
+    }
+
+    if (flags_copy)
+        OE_RAISE_MSG(OE_INVALID_PARAMETER, "invalid flag supplied");
+
+    if (length == 0)
+        OE_RAISE_MSG(OE_INVALID_PARAMETER, "length must be non zero");
+
+    // MAP_SHARED, MAP_SHARED_VALIDATE, MAP_PRIVATE are all treated as same
+    // since enclave is single process. Exactly one of them must be specified.
```

Comment:
```suggestion
    // MAP_SHARED, MAP_SHARED_VALIDATE, MAP_PRIVATE are all treated the same
    // since an enclave is a single process. Exactly one of them must be specified.
```

Reply:
Done.